src/HttpOverStream.Server.Owin/CustomListenerHost.cs

```diff
@@ -41,31 +41,40 @@ private async void OnAccept(Stream stream)
             {
                 using (stream)
                 {
+                    var sendHeadersCallback = new List<(Action<object>, object)>();
                     var owinContext = new OwinContext();
+                    // this is an owin extension
+                    owinContext.Set<Action<Action<object>, object>>("server.OnSendingHeaders", (callback, state) =>
+                    {
+                        sendHeadersCallback.Add((callback, state));
+                    });
                     owinContext.Set("owin.Version", "1.0");
+
                     Debug.WriteLine("Server: reading message..");
                     await PopulateRequestAsync(stream, owinContext.Request, CancellationToken.None).ConfigureAwait(false);
                     Debug.WriteLine("Server: finished reading message");
-                    using (var body = new MemoryStream())
+                    var body = new WriteInterceptStream(stream, async () =>

```

Comment:
I think this would be more readable if we extracted this delegate into a variable or a function to explain that its being triggered later